compiler/src/main/scala/org/scalajs/core/compiler/PreTyperComponent.scala

```diff
@@ -0,0 +1,123 @@
+ *    private def baz2 = ???
+ *  }
+ *  }}}
+ *
+ *  Would become:
+ *  {{{
+ *  class $anon extends ... {
+ *    @WasPublicBeforeTyper val foo = ???
+ *    @WasPublicBeforeTyper var bar = ???
+ *    @WasPublicBeforeTyper def baz = ???
+ *    private val foo2 = ???
+ *    private var bar2 = ???
+ *    private def baz2 = ???
+ *  }
+ *  }}}
+ *
+ *  And after `typer` (if has SI-9487) will be:
+ *  {{{
+ *  class $anon extends ... {
+ *    @WasPublicBeforeTyper private[this] var foo = ???
+ *    private <stable> <accessor> def foo = ??? // Needs fix
+ *
+ *    @WasPublicBeforeTyper private[this] var bar = ???
+ *    private <accessor> def bar = ??? // Needs fix
+ *    private <accessor> def bar_=(...) = ??? // Needs fix
+ *
+ *    @WasPublicBeforeTyper private def baz = ??? // Needs fix
+ *    ...
+ *  }
+ *  }}}
+ *
+ *  @author Nicolas Stucki
+ */
+abstract class PreTyperComponent
+    extends plugins.PluginComponent with transform.Transform with Compat210Component {
+
+  import global._
+
+  val phaseName = "jspretyper"
+
+  override protected def newTransformer(unit: CompilationUnit): Transformer =
+    new PreTyperTransformer(unit)
+
+  class PreTyperTransformer(unit: CompilationUnit) extends Transformer {
+    override def transform(tree: Tree): Tree = tree match {
+      case tree: ClassDef if needsAnnotations(tree) =>
+        val newBody = tree.impl.body.map {
+          case vdef: ValDef if needsAnnotations(vdef) =>
+            treeCopy.ValDef(vdef, withWasPublic(vdef.mods), vdef.name,
+                vdef.tpt, transform(vdef.rhs))
+
+          case ddef: DefDef if needsAnnotations(ddef) =>
+            treeCopy.DefDef(ddef, withWasPublic(ddef.mods), ddef.name,
+                ddef.tparams, ddef.vparamss, ddef.tpt, transform(ddef.rhs))
+
+          case member => transform(member)
+        }
+        val newImpl =
+          treeCopy.Template(tree.impl, tree.impl.parents, tree.impl.self, newBody)
+        treeCopy.ClassDef(tree, tree.mods, tree.name, tree.tparams, newImpl)
+
+      case _ => super.transform(tree)
+    }
+  }
+
+  private def needsAnnotations(classDef: ClassDef): Boolean = {
+    classDef.name == nme.ANON_CLASS_NAME.toTypeName &&
+    classDef.impl.body.exists {
+      case vdef: ValDef => needsAnnotations(vdef)
+      case ddef: DefDef => needsAnnotations(ddef)
+      case _ => false
+    }
+  }
+
+  private def needsAnnotations(vdef: ValDef): Boolean =
+    vdef.mods.isPublic
+
+  private def needsAnnotations(ddef: DefDef): Boolean =
+    ddef.mods.isPublic && ddef.name != nme.CONSTRUCTOR
+
+  private def withWasPublic(mods: Modifiers): Modifiers =
+    mods.withAnnotations(List(anonymousClassMethodWasPublicAnnotation))
+
+  private val scalajs = newTermName("scalajs")
+  private val wasPublicBeforeTyper = newTypeName("WasPublicBeforeTyper")
+
+  private lazy val anonymousClassMethodWasPublicAnnotation: Tree = {
+    val runtimePackage =
+      Select(Select(Ident(nme.scala_), scalajs), nme.runtime)
```

Comment:
Should prefix that even by a `Select(Ident(nme.ROOTPKG), ...`. Otherwise this can be broken if there is something named `scala` in scope.